server/aesdsocket.c

```diff
@@ -110,244 +133,421 @@ int main(int argc, char **argv)
+        }
+        syslog(LOG_INFO,"**** Wrote TS AESDSOCKET timer handler ****");
+    }
+    write_time = 0;
+    return 0;
+}
+
+/*
+* asesd_soc_timer_init 
+* This is ased timer init function that will
+* trigger at every 10 sec.
+* 
+* Parameters: None
+*    
+*
+* Returns: 0 for succcess and non-zero for error
+*/
+void asesd_soc_timer_handler(int signum) {
+    time_t now;
+    
+    syslog(LOG_INFO,"**** AESDSOCKET timer handler ****");
+    time_value[write_time++] = time(&now);
+    if (( write_time % MAX_TIME_ENTRY) == 0) {
+        write_time = 1;
+    }
+ 
+}
+
+/*
+* asesd_soc_timer_init 
+* This is ased timer init function that will
+* trigger at every 10 sec.
+* 
+* Parameters: None
+*    
+*
+* Returns: 0 for succcess and non-zero for error
+*/
+int  asesd_soc_timer_init(void)
+{
+    struct sigevent aesd_soc_tmr_event;
+    struct itimerspec aesd_soc_tmr_timer_spec;
+
+    aesd_soc_tmr_event.sigev_signo = SIGALRM;
+    aesd_soc_tmr_event.sigev_value.sival_ptr = &timerid;
+    aesd_soc_tmr_event.sigev_notify = SIGEV_SIGNAL;
+
+
+    if (timer_create(CLOCK_REALTIME, &aesd_soc_tmr_event, &timerid) != 0) {
+        syslog(LOG_INFO,"**** Unable to create timer ****");
+        return -1;
+    }
+
+    aesd_soc_tmr_timer_spec.it_value.tv_sec = 10;
+    aesd_soc_tmr_timer_spec.it_value.tv_nsec = 0;
+    aesd_soc_tmr_timer_spec.it_interval.tv_sec = 10; 
+    aesd_soc_tmr_timer_spec.it_interval.tv_nsec = 0; 
+
+
+    if (timer_settime(timerid, 0, &aesd_soc_tmr_timer_spec, NULL) != 0) {
+        syslog(LOG_INFO,"**** Unable to set the timer ****");
+        return -2;
+    }
+    
+    signal(SIGALRM, asesd_soc_timer_handler);
+
+    return 0;
+}
+
+
+/*
+* aesdsoc_sighandler
+* 
+* Parameters:
+*   signal_no:      Signal number. 
+*
+* Returns: None
+*/
+static void aesdsoc_sighandler(int signal_no) {
+
+    syslog(LOG_INFO, "Caught signal, exiting"); 
+    if (signal_no == SIGINT || signal_no == SIGTERM ) {
+       exit_aesd_soc = TRUE;
+       soc_close = TRUE;
+    }
+    if (signal_no == SIGINT )
+       file_close = FALSE;    
+}
+
+/*
+* aesdsocket thread clean-up function
+* 
+* Parameters: None
+*
+* Returns: 0 list is empty, otherwise 1; 
+*   
+*/
+int aesdsocket_memory_cleanup(void) {
+    ASEDSocThread_t *temp;
+    ASEDSocThread_t *ptr;
+    pthread_mutex_lock(&link_list_mutex);            
+    SLIST_FOREACH_SAFE(ptr, &head, entries, temp) {
+        SLIST_REMOVE(&head, ptr, ASEDSocThread, entries);
+        free(ptr);
+    }
+    
+    pthread_mutex_unlock(&link_list_mutex);
+    if (SLIST_EMPTY(&head)) {
+        return 0;
+    }
+    return 1;
+}
+
+/*
+* aesdsocket thread clean-up function
+* 
+* Parameters: None
+*
+* Returns: 0 list is empty, otherwise 1; 
+*   
+*/
+int aesdsocket_thread_cleanup(void) {
+    ASEDSocThread_t *temp;
+    ASEDSocThread_t *ptr;
+    pthread_mutex_lock(&link_list_mutex);            
+    SLIST_FOREACH_SAFE(ptr, &head, entries, temp) {
+        if (ptr->done == 1) {
```

Comment:
@sujoyray-ucolorado I think you'd need a version of this function which doesn't require the thread to be `done` in order to join, when cleaning up on program exit.  That way if a thread hadn't completed by the time you started cleanup you'd wait for it to exit before exiting the program.